homeassistant/components/control4/__init__.py

```diff
@@ -135,30 +98,161 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
-    return return_list
+    _LOGGER.debug("Getting items of category: %s", category)
+    director = hass.data[DOMAIN][entry.entry_id][CONF_DIRECTOR]
+    return_list = await director.getAllItemsByCategory(category)
+    return json.loads(return_list)
+
 
+async def refresh_tokens(hass: HomeAssistant, entry: ConfigEntry):
+    """Store updated authentication and director tokens in hass.data, and schedule next token refresh."""
+    config = entry.data
+    verify_ssl_session = aiohttp_client.async_get_clientsession(hass)
 
-class Control4Entity(CoordinatorEntity):
+    account = C4Account(
+        config[CONF_USERNAME], config[CONF_PASSWORD], verify_ssl_session
+    )
+    try:
+        await account.getAccountBearerToken()
+    except client_exceptions.ClientError as exception:
+        _LOGGER.error("Error connecting to Control4 account API: %s", exception)
+        raise ConfigEntryNotReady from exception
+    except BadCredentials as exception:
+        _LOGGER.error(
+            "Error authenticating with Control4 account API, incorrect username or password: %s",
+            exception,
+        )
+        raise ConfigEntryAuthFailed(exception) from exception
+
+    controller_unique_id = config[CONF_CONTROLLER_UNIQUE_ID]
+    director_token_dict = await account.getDirectorBearerToken(controller_unique_id)
+    no_verify_ssl_session = aiohttp_client.async_get_clientsession(
+        hass, verify_ssl=False
+    )
+
+    director = C4Director(
+        config[CONF_HOST], director_token_dict[CONF_TOKEN], no_verify_ssl_session
+    )
+
+    _LOGGER.debug("Saving new account and director tokens in hass data")
+    entry_data = hass.data[DOMAIN][entry.entry_id]
+    entry_data[CONF_ACCOUNT] = account
+    entry_data[CONF_DIRECTOR] = director
+
+    if not (
+        CONF_WEBSOCKET in entry_data
+        and isinstance(entry_data[CONF_WEBSOCKET], C4Websocket)
+    ):
+        _LOGGER.debug("First time setup, creating new C4Websocket object")
+        connection_tracker = C4WebsocketConnectionTracker(hass, entry)
+        websocket = C4Websocket(
+            config[CONF_HOST],
+            no_verify_ssl_session,
+            connection_tracker.connect_callback,
+            connection_tracker.disconnect_callback,
+        )
+        entry_data[CONF_WEBSOCKET] = websocket
+
+        # Silence C4Websocket related loggers, that would otherwise spam INFO logs with debugging messages
+        logging.getLogger("socketio.client").setLevel(logging.WARNING)
+        logging.getLogger("engineio.client").setLevel(logging.WARNING)
+        logging.getLogger("charset_normalizer").setLevel(logging.ERROR)
+
+    _LOGGER.debug("Starting new WebSocket connection")
+    await entry_data[CONF_WEBSOCKET].sio_connect(director.director_bearer_token)
+
+    _LOGGER.debug(
+        "Registering next token refresh in %s seconds",
+        director_token_dict["validSeconds"],
+    )
+    obj = RefreshTokensObject(hass, entry)
+    async_call_later(
```

Comment:
This returns a callable that can be used to cancel this.  Is it possible the integration can get unloaded before it runs?

Reply:
Yes it is possible for the integration to be unloaded before the scheduled task is run. I have updated the code to store the callback in `hass.data` and run it in `async_unload_entry`.